azurerm/internal/services/compute/virtual_machine.go

```diff
@@ -411,3 +416,407 @@ func flattenVirtualMachineOSDisk(ctx context.Context, disksClient *compute.Disks
+				"caching": {
+					Type:     schema.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(compute.CachingTypesNone),
+						string(compute.CachingTypesReadOnly),
+						string(compute.CachingTypesReadWrite),
+					}, false),
+				},
+
+				"lun": {
+					Type:         schema.TypeInt,
+					Required:     true,
+					ForceNew:     true,
+					ValidateFunc: validation.IntAtLeast(0),
+				},
+
+				"name": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+
+				"storage_account_type": {
+					Type:     schema.TypeString,
+					Required: true,
+					// whilst this appears in the Update block the API returns this when changing:
+					// Changing property 'osDisk.managedDisk.storageAccountType' is not allowed
+					ValidateFunc: validation.StringInSlice([]string{
+						// note: OS Disks don't support Ultra SSDs
+						string(compute.StorageAccountTypesPremiumLRS),
+						string(compute.StorageAccountTypesStandardLRS),
+						string(compute.StorageAccountTypesStandardSSDLRS),
+					}, false),
+				},
+
+				"disk_size_gb": {
+					Type:     schema.TypeInt,
+					Required: true,
+					// Max based on P80/S80/E80 Managed Disk type max size
+					ValidateFunc: validation.IntBetween(1, 32767),
+				},
+
+				"disk_encryption_set_id": {
+					Type:     schema.TypeString,
+					Optional: true,
+					// the Compute/VM API is broken and returns the Resource Group name in UPPERCASE
+					DiffSuppressFunc: suppress.CaseDifference,
+					ValidateFunc:     validate.DiskEncryptionSetID,
+				},
+
+				"write_accelerator_enabled": {
+					Type:     schema.TypeBool,
+					Optional: true,
+					Default:  false,
+				},
+			},
+		},
+		Set: resourceArmVirtualMachineNewDataDiskHash,
+	}
+}
+
+func virtualMachineExistingDataDiskSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeSet,
+		Optional: true,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"managed_disk_id": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validate.ManagedDiskID,
+				},
+
+				"lun": {
+					Type:         schema.TypeInt,
+					Required:     true,
+					ForceNew:     true,
+					ValidateFunc: validation.IntAtLeast(0),
+				},
+
+				"caching": {Type: schema.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						string(compute.CachingTypesNone),
+						string(compute.CachingTypesReadOnly),
+						string(compute.CachingTypesReadWrite),
+					}, false),
+				},
+
+				"storage_account_type": {
+					Type:     schema.TypeString,
+					Required: true,
+					// whilst this appears in the Update block the API returns this when changing:
+					// Changing property 'osDisk.managedDisk.storageAccountType' is not allowed
+					ValidateFunc: validation.StringInSlice([]string{
+						// note: OS Disks don't support Ultra SSDs
```

Comment:
(as above) the comment isn't valid for Data Disks and I think we should support Ultra disks here?